temporal-sdk/src/main/java/io/temporal/internal/common/SearchAttributePayloadConverter.java

```diff
@@ -46,6 +47,84 @@ final class SearchAttributePayloadConverter {
   public static final SearchAttributePayloadConverter INSTANCE =
       new SearchAttributePayloadConverter();
 
+  public Payload encodeTyped(SearchAttributeKey<?> key, @Nullable Object value) {
+    // We can encode as-is because we know it's strictly typed to expected key value. We
+    // accept a null value because updates for unset can be null.
+    return DefaultDataConverter.STANDARD_INSTANCE.toPayload(value).get().toBuilder()
+        .putMetadata(
+            METADATA_TYPE_KEY,
+            ByteString.copyFromUtf8(indexValueTypeToEncodedValue(key.getValueType())))
+        .build();
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  public void decodeTyped(SearchAttributes.Builder builder, String name, @Nonnull Payload payload) {
+    // Get key type
+    SearchAttributeKey key;
+    IndexedValueType indexType = getIndexType(payload.getMetadataMap().get(METADATA_TYPE_KEY));
+    switch (indexType) {
+      case INDEXED_VALUE_TYPE_TEXT:
+        key = SearchAttributeKey.forText(name);
+        break;
+      case INDEXED_VALUE_TYPE_KEYWORD:
+        key = SearchAttributeKey.forKeyword(name);
+        break;
+      case INDEXED_VALUE_TYPE_INT:
+        key = SearchAttributeKey.forLong(name);
+        break;
+      case INDEXED_VALUE_TYPE_DOUBLE:
+        key = SearchAttributeKey.forDouble(name);
+        break;
+      case INDEXED_VALUE_TYPE_BOOL:
+        key = SearchAttributeKey.forBoolean(name);
+        break;
+      case INDEXED_VALUE_TYPE_DATETIME:
+        key = SearchAttributeKey.forOffsetDateTime(name);
+        break;
+      case INDEXED_VALUE_TYPE_KEYWORD_LIST:
+        key = SearchAttributeKey.forKeywordList(name);
+        break;
+      default:
+        log.warn(
+            "[BUG] Unrecognized indexed value type {} on search attribute key {}", indexType, name);
+        return;
+    }
+
+    // Attempt conversion to key type or if not keyword list, single-value list and extract out
+    try {
+      Object value =
```

Comment:
What happens if there are multiple values in the payload or the payload is an array with a single item and the typed attribute is not a list?
Seems like we should log a warning and (possibly) put the first value from the array in the attribute, at least for single item arrays.
In any case, just verifying this won't fail WFTs.
Also acceptable if we warn and users have to use untyped search attributes to access the values.

Reply:
> It will, by intention. If you have chosen to move to typed search attributes, you have chosen the strict guarantees that come with it (and that come with our new SA validations we do now). It is only possible to violate these if you used non-typed search attribute API (and advanced self-hosted validation IIUC) which I don't believe needs to lead to silent truncate behavior.

What if a client from another language or SDK version starts a workflow with extra values in a keyword array?
The server accepts this but workflows will get stuck.

Reply:
> What if a client from another language or SDK version starts a workflow with extra values in a keyword array?

They chose to mix non-typed-search-attribute API usage with strict typed-search-attribute usage. Yes I can understand the concern about half of a user's code using non-strict and the other half using strict, but we want typed search attributes to be strict here. Untyped is not "all value types", untyped is "all values".

Reply:
Okay, I buy that, we'll just have to explain this limitation to users and how to fix get their workflows "unstuck".

Reply:
:+1: Limitation accepted for using deprecated API